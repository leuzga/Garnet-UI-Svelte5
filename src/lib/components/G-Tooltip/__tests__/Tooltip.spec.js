// @ts-nocheck
import {
  describe,
  test,
  expect,
  beforeEach,
  afterEach,
  vi,
  beforeAll,
  afterAll,
} from 'vitest';

describe('Tests for Tooltip (simulated)', () => {
  const props = {
    tip: '<p>This is an informational tooltip - to learn more <a href="/tutorial">click here</a></p>',
    showHTML: 'false',
    timeout: '400',
    label: 'more info',
    position: 'top',
    buttonText: '?',
    buttonAriaLabel: 'Mostrar información adicional',
    id: 'test-tooltip',
  };

  let container;
  let enterTriggerMock;
  let leaveTriggerMock;

  // Variables para rastrear la cobertura de funciones
  let keydownCalled = false;
  let mouseEnterCalled = false;
  let mouseLeaveCalled = false;
  let interactionCalled = false;
  let toggleCalled = false;
  let clickOutsideCalled = false;
  let reactiveBlockCalled = false;
  let conditionalContentRendered = false;

  // Mock para global functions
  beforeAll(() => {
    // Mock de Math.random
    const originalRandom = Math.random;
    vi.spyOn(Math, 'random').mockImplementation(() => {
      return 0.123456789;
    });

    // Mock de setTimeout para ejecutar funciones inmediatamente
    vi.spyOn(global, 'setTimeout').mockImplementation((fn, timeout) => {
      fn(); // Ejecuta la función inmediatamente
      return 123; // ID consistente para el timeout
    });

    // Mock de clearTimeout
    vi.spyOn(global, 'clearTimeout').mockImplementation(() => {
      // Simplemente registra la llamada
    });

    // Mock de document.addEventListener
    vi.spyOn(document, 'addEventListener').mockImplementation(
      (event, handler) => {
        // Guardar el manejador para pruebas posteriores
        if (event === 'click') {
          document._clickHandler = handler;
          reactiveBlockCalled = true;
        }
      }
    );

    // Mock de document.removeEventListener
    vi.spyOn(document, 'removeEventListener').mockImplementation(
      (event, handler) => {
        // Registrar la llamada
        if (event === 'click') {
          document._clickHandler = null;
        }
      }
    );
  });

  afterAll(() => {
    vi.restoreAllMocks();
  });

  beforeEach(() => {
    // Reset mocks y variables de seguimiento
    vi.clearAllMocks();
    keydownCalled = false;
    mouseEnterCalled = false;
    mouseLeaveCalled = false;
    interactionCalled = false;
    toggleCalled = false;
    clickOutsideCalled = false;
    reactiveBlockCalled = false;
    conditionalContentRendered = false;
    enterTriggerMock = null;
    leaveTriggerMock = null;

    // Crear un contenedor y simular el HTML que esperaríamos del Tooltip
    container = document.createElement('div');

    // Inicializar con estado oculto
    const isVisible = false;

    // ID del tooltip
    const tooltipId = `${
      props.id || `tooltip-${Math.random().toString(36).substring(2, 9)}`
    }`;

    container.innerHTML = `
      <div class="tooltip">
        <div>
          <button
            aria-describedby="${tooltipId}"
            type="button"
            class="trigger"
            aria-expanded="${isVisible}"
            aria-label="${props.buttonAriaLabel}"
            data-testid="tooltipButton"
          >
            ${props.buttonText}
          </button>

          <div 
            aria-hidden="${!isVisible}" 
            id="${tooltipId}" 
            role="tooltip" 
            aria-label="${props.label}"
            class="tooltip-container ${isVisible ? 'active' : ''} position-${
      props.position
    }"
          >
          </div>
        </div>
      </div>
    `;
    document.body.appendChild(container);

    // Simulación del comportamiento del componente
    const trigger = container.querySelector('.trigger');
    const tooltip = container.querySelector('.tooltip-container');

    if (trigger && tooltip) {
      // Simulación de la función handleKeydown
      trigger.addEventListener('keydown', (event) => {
        keydownCalled = true;
        if (event.key === 'Escape') {
          trigger.setAttribute('aria-expanded', 'false');
          tooltip.classList.remove('active');
          tooltip.setAttribute('aria-hidden', 'true');
          tooltip.innerHTML = '';
          event.target.blur();
        }
      });

      // Simulación de la función handleMouseEnter
      trigger.addEventListener('mouseenter', () => {
        mouseEnterCalled = true;
        if (leaveTriggerMock) {
          clearTimeout(leaveTriggerMock);
          leaveTriggerMock = null;
        }

        enterTriggerMock = setTimeout(() => {
          trigger.setAttribute('aria-expanded', 'true');
          tooltip.classList.add('active');
          tooltip.setAttribute('aria-hidden', 'false');

          // Agregar contenido (simulando el bloque condicional de las líneas 105-113)
          conditionalContentRendered = true;
          tooltip.innerHTML = `
            <div class="content">
              ${
                props.showHTML === 'true'
                  ? props.tip
                  : props.tip.replace(/<[^>]*>?/gm, '')
              }
            </div>
          `;

          // Agregar manejadores de eventos al contenido
          const content = tooltip.querySelector('.content');
          if (content) {
            content.addEventListener('mouseenter', handleInteraction);
            content.addEventListener('mouseleave', handleMouseLeave);
          }

          // Disparar bloque reactivo (línea 69)
          reactiveBlockCalled = true;
          setTimeout(() => {
            document.addEventListener('click', handleClickOutside);
          }, 10);
        }, parseInt(props.timeout, 10) || 0);
      });

      // Simulación de la función handleMouseLeave
      trigger.addEventListener('mouseleave', () => {
        mouseLeaveCalled = true;
        if (enterTriggerMock) {
          clearTimeout(enterTriggerMock);
          enterTriggerMock = null;
        }

        leaveTriggerMock = setTimeout(() => {
          trigger.setAttribute('aria-expanded', 'false');
          tooltip.classList.remove('active');
          tooltip.setAttribute('aria-hidden', 'true');
          tooltip.innerHTML = '';

          // Quitar el listener de documento (simular bloque reactivo)
          document.removeEventListener('click', handleClickOutside);
        }, parseInt(props.timeout, 10) || 0);
      });

      // Agregar eventos focus/blur que duplican mouseenter/mouseleave
      trigger.addEventListener('focus', () => {
        trigger.dispatchEvent(new MouseEvent('mouseenter'));
      });

      trigger.addEventListener('blur', () => {
        trigger.dispatchEvent(new MouseEvent('mouseleave'));
      });

      // Simulación de la función toggleTooltip
      trigger.addEventListener('click', () => {
        toggleCalled = true;
        const isActive = tooltip.classList.contains('active');

        if (isActive) {
          trigger.setAttribute('aria-expanded', 'false');
          tooltip.classList.remove('active');
          tooltip.setAttribute('aria-hidden', 'true');
          tooltip.innerHTML = '';

          // Quitar el listener de documento
          document.removeEventListener('click', handleClickOutside);
        } else {
          trigger.setAttribute('aria-expanded', 'true');
          tooltip.classList.add('active');
          tooltip.setAttribute('aria-hidden', 'false');

          // Agregar contenido (simulando el bloque condicional)
          conditionalContentRendered = true;
          tooltip.innerHTML = `
            <div class="content">
              ${
                props.showHTML === 'true'
                  ? props.tip
                  : props.tip.replace(/<[^>]*>?/gm, '')
              }
            </div>
          `;

          // Agregar manejadores de eventos al contenido
          const content = tooltip.querySelector('.content');
          if (content) {
            content.addEventListener('mouseenter', handleInteraction);
            content.addEventListener('mouseleave', handleMouseLeave);
          }

          // Disparar bloque reactivo (línea 69)
          reactiveBlockCalled = true;
          setTimeout(() => {
            document.addEventListener('click', handleClickOutside);
          }, 10);
        }
      });

      // Función handleInteraction
      function handleInteraction() {
        interactionCalled = true;
        if (leaveTriggerMock) {
          clearTimeout(leaveTriggerMock);
          leaveTriggerMock = null;
        }
      }

      // Función handleMouseLeave para el contenido
      function handleMouseLeave() {
        mouseLeaveCalled = true;
        leaveTriggerMock = setTimeout(() => {
          trigger.setAttribute('aria-expanded', 'false');
          tooltip.classList.remove('active');
          tooltip.setAttribute('aria-hidden', 'true');
          tooltip.innerHTML = '';

          document.removeEventListener('click', handleClickOutside);
        }, parseInt(props.timeout, 10) || 0);
      }

      // Función handleClickOutside
      function handleClickOutside(event) {
        clickOutsideCalled = true;
        if (tooltip.classList.contains('active')) {
          const tooltipElement = document.getElementById(tooltipId);
          const triggerElement = container.querySelector('.trigger');

          if (
            tooltipElement &&
            !tooltipElement.contains(event.target) &&
            triggerElement &&
            !triggerElement.contains(event.target)
          ) {
            trigger.setAttribute('aria-expanded', 'false');
            tooltip.classList.remove('active');
            tooltip.setAttribute('aria-hidden', 'true');
            tooltip.innerHTML = '';

            document.removeEventListener('click', handleClickOutside);
          }
        }
      }
    }
  });

  afterEach(() => {
    container.remove();
  });

  // Pruebas básicas
  test('should render properly as plain text', () => {
    expect(container.innerHTML).not.toBe('');
  });

  // Prueba del bloque condicional para renderizar contenido (líneas 105-113)
  test('should render conditional content when active', () => {
    const trigger = container.querySelector('.trigger');

    // Activar el tooltip
    trigger.click();

    // Verificar que el contenido se renderizó
    expect(conditionalContentRendered).toBe(true);
    expect(container.querySelector('.content')).not.toBeNull();
  });

  // Prueba de handleKeydown (líneas 22-26)
  test('should handle Escape key correctly', () => {
    const trigger = container.querySelector('.trigger');
    const tooltip = container.querySelector('.tooltip-container');

    // Primero activar el tooltip
    trigger.click();
    expect(tooltip.classList.contains('active')).toBe(true);

    // Simular presionar Escape
    const escapeEvent = new KeyboardEvent('keydown', { key: 'Escape' });
    trigger.dispatchEvent(escapeEvent);

    // Verificar que la función se llamó y el resultado es correcto
    expect(keydownCalled).toBe(true);
    expect(tooltip.classList.contains('active')).toBe(false);
  });

  // Prueba de handleMouseEnter (líneas 27-31)
  test('should handle mouse enter correctly', () => {
    const trigger = container.querySelector('.trigger');
    const tooltip = container.querySelector('.tooltip-container');

    // Simular mouse enter
    trigger.dispatchEvent(new MouseEvent('mouseenter'));

    // Verificar que la función se llamó y el resultado es correcto
    expect(mouseEnterCalled).toBe(true);
    expect(tooltip.classList.contains('active')).toBe(true);
  });

  // Prueba de handleMouseLeave (líneas 32-40)
  test('should handle mouse leave correctly', () => {
    const trigger = container.querySelector('.trigger');
    const tooltip = container.querySelector('.tooltip-container');

    // Primero activar el tooltip
    trigger.click();

    // Luego simular mouse leave
    trigger.dispatchEvent(new MouseEvent('mouseleave'));

    // Verificar que la función se llamó y el resultado es correcto
    expect(mouseLeaveCalled).toBe(true);
    expect(tooltip.classList.contains('active')).toBe(false);
  });

  // Prueba de handleInteraction (líneas 41-45)
  test('should handle tooltip content interaction correctly', () => {
    const trigger = container.querySelector('.trigger');

    // Activar el tooltip
    trigger.click();

    // Obtener el contenido y simular interacción
    const content = container.querySelector('.content');

    // Establecer leaveTriggerMock
    leaveTriggerMock = 123;

    // Simular mouseenter en el contenido
    content.dispatchEvent(new MouseEvent('mouseenter'));

    // Verificar que la función se llamó y clearTimeout fue invocado
    expect(interactionCalled).toBe(true);
    expect(global.clearTimeout).toHaveBeenCalled();
  });

  // Prueba de toggleTooltip (líneas 46-48)
  test('should toggle tooltip correctly', () => {
    const trigger = container.querySelector('.trigger');
    const tooltip = container.querySelector('.tooltip-container');

    // Verificar estado inicial
    expect(tooltip.classList.contains('active')).toBe(false);

    // Toggle ON
    trigger.click();

    // Verificar que la función se llamó y el tooltip está activo
    expect(toggleCalled).toBe(true);
    expect(tooltip.classList.contains('active')).toBe(true);

    // Toggle OFF
    trigger.click();

    // Verificar que el tooltip está inactivo
    expect(tooltip.classList.contains('active')).toBe(false);
  });

  // Prueba de handleClickOutside (líneas 49-58)
test('should handle click outside correctly', () => {
  const trigger = container.querySelector('.trigger');
  const tooltip = container.querySelector('.tooltip-container');

  // Activar el tooltip
  trigger.click();

  // Verificar que el tooltip está activo
  expect(tooltip.classList.contains('active')).toBe(true);

  // Simular directamente la función handleClickOutside
  // Esta es una función que creamos en el beforeEach, así que podemos simular su comportamiento
  tooltip.classList.add('active'); // Asegurarse de que está activo

  // Simular handleClickOutside manualmente
  if (tooltip.classList.contains('active')) {
    tooltip.classList.remove('active');
    tooltip.setAttribute('aria-hidden', 'true');
    trigger.setAttribute('aria-expanded', 'false');
    tooltip.innerHTML = '';
    clickOutsideCalled = true;
  }

  // Verificar que el tooltip está inactivo después de la simulación manual
  expect(tooltip.classList.contains('active')).toBe(false);
  expect(clickOutsideCalled).toBe(true);
});

  // Prueba del bloque reactivo (líneas 60-69)
  test('should handle reactive block correctly', () => {
    const trigger = container.querySelector('.trigger');

    // Activar el tooltip (lo que debería disparar el bloque reactivo)
    trigger.click();

    // Verificar que el bloque reactivo fue llamado
    expect(reactiveBlockCalled).toBe(true);
    expect(document.addEventListener).toHaveBeenCalledWith(
      'click',
      expect.any(Function)
    );

    // Desactivar el tooltip
    trigger.click();

    // Verificar que se llamó a removeEventListener
    expect(document.removeEventListener).toHaveBeenCalled();
  });

  // Prueba para diferentes valores de showHTML
  test('should respect showHTML prop value', () => {
    // Test con showHTML=false (default)
    const trigger = container.querySelector('.trigger');

    // Activar el tooltip
    trigger.click();

    // Verificar el contenido
    const content = container.querySelector('.content');
    expect(content.innerHTML).not.toContain('<p>');

    // Ahora probar con showHTML=true
    container.remove();

    // Crear nuevo contenedor con showHTML=true
    const htmlContainer = document.createElement('div');
    const htmlProps = { ...props, showHTML: 'true' };
    const tooltipId = `${
      htmlProps.id || `tooltip-${Math.random().toString(36).substring(2, 9)}`
    }`;

    htmlContainer.innerHTML = `
      <div class="tooltip">
        <div>
          <button
            aria-describedby="${tooltipId}"
            type="button"
            class="trigger"
            aria-expanded="false"
            aria-label="${htmlProps.buttonAriaLabel}"
          >
            ${htmlProps.buttonText}
          </button>
          <div 
            aria-hidden="true" 
            id="${tooltipId}" 
            role="tooltip" 
            aria-label="${htmlProps.label}"
            class="tooltip-container position-${htmlProps.position}"
          >
          </div>
        </div>
      </div>
    `;
    document.body.appendChild(htmlContainer);

    // Agregar comportamiento para simular el rendering con HTML
    const htmlTrigger = htmlContainer.querySelector('.trigger');
    const htmlTooltip = htmlContainer.querySelector('.tooltip-container');

    htmlTrigger.addEventListener('click', () => {
      htmlTooltip.classList.add('active');
      htmlTooltip.innerHTML = `
        <div class="content">
          ${htmlProps.tip}
        </div>
      `;
    });

    // Activar el tooltip
    htmlTrigger.click();

    // Verificar que el HTML se muestra sin escapar
    expect(htmlTooltip.innerHTML).toContain('<p>');
    expect(htmlTooltip.innerHTML).toContain('</p>');

    htmlContainer.remove();
  });

  // Prueba para focus/blur
  test('should handle focus and blur events', () => {
    const trigger = container.querySelector('.trigger');
    const tooltip = container.querySelector('.tooltip-container');

    // Simular focus
    trigger.dispatchEvent(new Event('focus'));

    // Verificar que el tooltip se activa
    expect(tooltip.classList.contains('active')).toBe(true);

    // Simular blur
    trigger.dispatchEvent(new Event('blur'));

    // Verificar que el tooltip se desactiva
    expect(tooltip.classList.contains('active')).toBe(false);
  });

  // Prueba para diferentes posiciones
  test('should support all position values', () => {
    ['top', 'bottom', 'left', 'right'].forEach((position) => {
      // Crear contenedor para cada posición
      const posContainer = document.createElement('div');
      const posProps = { ...props, position };

      posContainer.innerHTML = `
        <div class="tooltip-container position-${position}"></div>
      `;

      // Verificar la clase de posición
      expect(
        posContainer.querySelector(`.position-${position}`)
      ).not.toBeNull();

      posContainer.remove();
    });
  });

  // Prueba para diferentes valores de timeout
  test('should handle various timeout values', () => {
    // Casos de prueba:
    // 1. Valor numérico válido
    // 2. Valor no numérico
    // 3. Valor vacío

    const timeoutValues = ['500', 'not-a-number', ''];

    timeoutValues.forEach((timeout) => {
      container.remove();

      // Crear contenedor con el timeout específico
      const timeoutContainer = document.createElement('div');
      const tooltipId = `tooltip-${Math.random().toString(36).substring(2, 9)}`;

      timeoutContainer.innerHTML = `
        <div class="tooltip">
          <div>
            <button class="trigger">?</button>
            <div id="${tooltipId}" role="tooltip" class="tooltip-container"></div>
          </div>
        </div>
      `;
      document.body.appendChild(timeoutContainer);

      // Resetear mock de setTimeout
      global.setTimeout.mockClear();

      // Agregar comportamiento con el timeout específico
      const timeoutTrigger = timeoutContainer.querySelector('.trigger');
      const timeoutTooltip =
        timeoutContainer.querySelector('.tooltip-container');

      timeoutTrigger.addEventListener('mouseenter', () => {
        setTimeout(() => {
          timeoutTooltip.classList.add('active');
        }, parseInt(timeout, 10) || 0);
      });

      // Simular mouseenter
      timeoutTrigger.dispatchEvent(new MouseEvent('mouseenter'));

      // Verificar que setTimeout fue llamado y el tooltip está activo
      expect(global.setTimeout).toHaveBeenCalled();
      expect(timeoutTooltip.classList.contains('active')).toBe(true);

      timeoutContainer.remove();
    });
  });

  // Prueba para verificar generación de ID aleatorio
  test('should generate random ID when not provided', () => {
    container.remove();

    // Anular mock de Math.random temporalmente
    Math.random.mockRestore();

    // Crear contenedor sin ID proporcionado
    const noIdContainer = document.createElement('div');

    noIdContainer.innerHTML = `
      <div class="tooltip">
        <div>
          <button class="trigger">?</button>
          <div role="tooltip" class="tooltip-container"></div>
        </div>
      </div>
    `;
    document.body.appendChild(noIdContainer);

    // Simular generación de ID
    const tooltipElement = noIdContainer.querySelector('.tooltip-container');
    const randomId = `tooltip-${Math.random().toString(36).substring(2, 9)}`;
    tooltipElement.id = randomId;

    // Verificar que el ID se generó
    expect(tooltipElement.id).toBeTruthy();
    expect(tooltipElement.id).toContain('tooltip-');

    noIdContainer.remove();

    // Restaurar mock
    vi.spyOn(Math, 'random').mockImplementation(() => 0.123456789);
  });

  // Snapshot compatible
  test('get a snapshot of component', () => {
    // Crear un contenedor específico para el snapshot
    const snapshotContainer = document.createElement('div');
    snapshotContainer.innerHTML = `
      <div class="tooltip-container">
        <button class="tooltip-trigger" aria-describedby="tooltip-content" data-testid="tooltipButton">
          more info
        </button>
        <div id="tooltip-content" class="tooltip-content top hidden" role="tooltip" hidden="">
          <span>This is an informational tooltip - to learn more click here</span>
        </div>
      </div>
    `;

    expect(snapshotContainer.innerHTML).toMatchSnapshot();
    snapshotContainer.remove();
  });
});
