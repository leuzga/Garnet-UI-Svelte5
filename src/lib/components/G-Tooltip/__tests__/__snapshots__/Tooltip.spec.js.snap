// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for Tooltip (simulated) > get a snapshot of component 1`] = `
"
      <div class="tooltip-container">
        <button class="tooltip-trigger" aria-describedby="tooltip-content" data-testid="tooltipButton">
          more info
        </button>
        <div id="tooltip-content" class="tooltip-content top hidden" role="tooltip" hidden="">
          <span>This is an informational tooltip - to learn more click here</span>
        </div>
      </div>
    "
`;
